### ISSUE TITLE
Multi-cluster: Adequately find incoming traffic from other clusters (multiple control plane)

### ISSUE THREAD

**Issue by israel-hdez** [kiali/kiali#1480]
In a simple scenario when a workload is calling a service in a remote cluster. For example this is the relevant part of the graph in the source cluster:

![image](https://user-images.githubusercontent.com/23639005/63721211-67557400-c816-11e9-862a-9b60c61bcee4.png)

The traffic generator is making requests to the httpbin-bar ServiceEntry, which is a service in a remote cluster.

The graph in the remote cluster is this one:

![image](https://user-images.githubusercontent.com/23639005/63721289-8c49e700-c816-11e9-8bdc-78c8f6db3228.png)

The "Unused nodes" option is enabled to show the issue. It is possible to see that the graph shows no traffic. However, the side panel is rightly stating that there is some traffic coming to the bar namespace.

Strangely, the Traffic tab (of the remote cluster) of the httpbin workload shows this:

![image](https://user-images.githubusercontent.com/23639005/63721685-5822f600-c817-11e9-9270-540efbebb537.png)

This means that telemetry is in place and, probably, data is being returned correctly. Probably, it's the graph the one that is not correctly displaying the edges.

**Comment by israel-hdez** [comment]
Telemetry for the cluster being destination of traffic:

```
istio_requests_total{
  connection_security_policy="mutual_tls",
  destination_app="httpbin",
  destination_principal="cluster.local/ns/bar/sa/default",
  destination_service="httpbin.bar.global",
  destination_service_name="httpbin.bar.global",
  destination_service_namespace="default",
  destination_version="v1",
  destination_workload="httpbin",
  destination_workload_namespace="bar",
  instance="172.17.0.18:42422",
  job="istio-mesh",
  permissive_response_code="none",
  permissive_response_policyid="none",
  reporter="destination",
  request_protocol="http",
  response_code="200",
  response_flags="-",
  source_app="unknown",
  source_principal="cluster.local/ns/bookinfo/sa/default",
  source_version="unknown",
  source_workload="unknown",
  source_workload_namespace="unknown"}	
```

Curiously, telemetry has destination_service_namespace="default", instead of "bar" value.
Apparently, the "default" is coming from the source cluster. In the source cluster, the ServiceEntry that routes traffic to the remote cluster is under the _default_ namespace. I'll move that ServiceEntry of the source cluster to another namespace to see if telemetry changes.

**Comment by jshaughn** [comment]
I guess I see what it is doing, but it doesn't seem right.  On ClusterA a source workload requests the httpbin service.  On ClusterA the httpbin service is a host on the serviceEntry defined in the Default ns.  The serviceEntry then sends the request to ClusterB (in this case not out of the mesh, but inside the mesh to another cluster).  On ClusterA we should see telemetry from the source workload to the serviceEntry (or possibly the PassthroughCluster in 1.3).  On ClusterB I'd expect telemetry from "unknown" to the httpbin workload.  But I think the destination_service_namespace should be with respect to ClusterB, and be "bar", because that is where the httpbin service is defined on ClusterB.

I have never totally understood why there is a distinction between destination_service_namespace and destination_workload_namespace, and I still don't think it should differ in this case.  We should probably confirm with Istio as to whether this is a bug, or intentional.  If intentional we may need to re-visit our queries.

**Comment by israel-hdez** [comment]
So, yes... In the source cluster, changing the namespace of the ServiceEntry has impact on the telemetry of the remote cluster. I moved it from the _default_ to the _test_ namespace, and there is a new timeserie (on ClusterB, the destination):

```
istio_requests_total{
  connection_security_policy="mutual_tls",
  destination_app="httpbin",
  destination_principal="cluster.local/ns/bar/sa/default",
  destination_service="httpbin.bar.global",
  destination_service_name="httpbin.bar.global",
  destination_service_namespace="test",
  destination_version="v1",
  destination_workload="httpbin",
  destination_workload_namespace="bar",
  instance="172.17.0.18:42422",
  job="istio-mesh",
  permissive_response_code="none",
  permissive_response_policyid="none",
  reporter="destination",
  request_protocol="http",
  response_code="200",
  response_flags="-",
  source_app="unknown",
  source_principal="cluster.local/ns/bookinfo/sa/default",
  source_version="unknown",
  source_workload="unknown",
  source_workload_namespace="unknown"}	
```

Note that only the destination_service_namespace label has changed.

This means that we cannot rely on the _destination_service_namespace_ label to query time series belonging to traffic that is incoming from a remote cluster.

**Comment by israel-hdez** [comment]
@jshaughn Yes, I will open a thread on Istio's discuss to see what they say.
It is too weird that destination_service_namespace records the value from "ClusterA".

**Comment by israel-hdez** [comment]
Question in Istio's discuss: https://discuss.istio.io/t/multi-cluster-destination-service-labels/3694

**Comment by israel-hdez** [comment]
An IRC discussion with @jshaughn pointed to conclude that the problematic code are queries to prometheus in the buildNamespaceTrafficMap of the `graph/telemetry/istio/istio.go`. Queries like this one

https://github.com/kiali/kiali/blob/d97ee0fbcfafc2d3b80c9d2e43667d7ae1a88c99/graph/telemetry/istio/istio.go#L261

where timeseries are filtered using the destination_service_namespace. Since the destination cluster (ClusterB) this label has "test" value (which is unexpected for Kiali), timeseries are being filtered out.

I found the PR that first introduced the queries that we currently have: https://github.com/kiali/kiali/pull/361. No comments about why using destination_service_namespace for the prometheus queries instead of destination_workload_namespace.

**Comment by israel-hdez** [comment]
Thinking a little about it, perhaps _destination_service_namespace_ is safer in general. So, to solve this issue, it's not as easy as just changing the queries to using destination_workload_namespace:
* Queries to ServiceEntries usually have destination_workload_namespace="unknown"
  * Even in the multi-cluster case, this happens. See the last time-serie at my comment: https://github.com/kiali/kiali/issues/1364#issuecomment-524434279
* Traffic to non-existent hosts may have destination_workload_namespace="unknown"
* Wierdly, workload-to-workload traffic populates destination_service_namespace with a value (see test: https://github.com/kiali/kiali/blob/d97ee0fbcfafc2d3b80c9d2e43667d7ae1a88c99/handlers/graph_test.go#L184-L197)

If Istio people confirms that the telemetry is as intended, I would lean towards using a dedicated query for the multi-cluster case. If it's an Istio bug, we should just wait for a fix.

**Comment by jshaughn** [comment]
@israel-hdez , I think the current queries are correct because not every request reaches a destination workload.  In general there will be a destination_service, although I don't think it's guaranteed; remember that the destination service may not be the requested service, due to virtual service manipulation.

So, if this is not considered a bug in Istio and is not fixed, then we unfortunately will probably need to look at another query.

**Comment by israel-hdez** [comment]
I got a reply in Istio's discuss. They say that current behavior is the expected one.

They added something important to keep in mind when building queries:

> The destination.service.* attributes are actually added by the client-side proxy, and forwarded to the server-side proxy for use in policy/telemetry. This is a bit counter-intuitive, but it reflects the fact that the destination workload does not know any information about the service through which it was invoked.

I guess this means a new query will be needed in Kiali to handle correctly multi-cluster traffic at "ClusterB".

**Comment by israel-hdez** [comment]
Fixed with #1651 .
